documentation/groff/gallery.1
.\" Author Ben Lavery
.\" Contact the author: ben.lavery@gmail.com
.\" Created on 29/05/2010
.\" Last edited on 30/05/2010
.\" Revision 3
.\" 
.\" LICENCE:
.\" This is licensed under the Creative Commons Licence.
.\" To view the human-readable summary of this licence, please direct your browser to:
.\" http://creativecommons.org/licenses/by-sa/2.0/uk/
.\" For the full licence, direct your browser to:
.\" http://creativecommons.org/licenses/by-sa/2.0/uk/legalcode
.\"
.\" Format this file with
.\" groff -man -Tascii gallery.1
.\" Generate PS file with
.\" groff -t -e -mandoc -Tps gallery.1 > gallery.ps
.\" Generate HTML file with
.\" groff -man -Tascii gallery.1 | man2html -title '<Gallery>' > gallery.html
.\"
.TH GALLERY 1 "MAY 2010" "UNIX" "User Manuals"
.SH NAME
gallery \- Generates a gallery of images.
.SH SYNOPSIS
.BR gallery " [-" ChHqTV "] [-" l " | -" L "] [-" "B big_image_size_in_pixels" "] [-" "c number_of_columns" "] [-" "D description_location" "] [-" "e extension" "] [-" "H html_version" "] [-" "t title" "] [-" "r relative_to" "] [-" "S small_image_size_in_pixels" "] [-" "s source_directory" "] -" "d directory"
.SH DESCRIPTION
The
.B gallery
command will generate a gallery of images.  Pass it a directory full of images and it will resize them and format an HTML file to display the images.  From revision 47 you can pass it a source directory and it will copy the contents of the directory to the directory specified under the -d option.

The script creates two directories in the directory passed to it:
.I bigs
which stores the larger images for the gallery (longest edge is 640px) and
.I thumbs
which contain thumbnails of the images (longest edge is 128px).  The script uses
.BR sips (1)
in OS X to do the image manipulation and a combination of
.BR convert "(1) and " identify (1)
in other UNIX environments.

After resizing, the script formats a file called
.IB index. EXTENSION
.RI ( index.php
by default).

The script uses some specific CSS
.I id
and
.I class
attributes.  These attributes are as follows:
.IP "id='gallery-body'"
This is a div that encompasses the entire gallery
.IP "class='gallery-column'"
This is a div which encompasses a
.B ROW
of images.  By default there are four images per 'gallery-column', but the user may specify otherwise.  This is placed inside the 'gallery-body' div.
.IP "class='gallery-image'"
This is a div that encompasses a single image and is placed inside the 'gallery-column' div.
.IP "class='gallery-caption'"
This is a paragraph which includes the caption text of your image.  This is placed inside the 'gallery-image' div, underneath the image.

.SH OPTIONS
.IP "-B big_image_size_in_pixels"
The size of the main images.  The number is applied to the longest side and is measured in pixels.
.IP "-c number of columns"
Number of columns per row.  Default is four.
.IP "-C"
Use captions \- These are generated using the file names without the file extension.
.IP "-d directory"
Absolute link to directory with images in.  Escape spaces with a \ symbol.
.IP "-D description_location"
Add the description found in
.I description.txt
which should be located in the directory specified with -d.
.IP -e
Extension for the gallery file.  Default is php.
.IP -h
Displays help text.
.IP "-H html_version"
Create a standalone HTML page.  Specify HTML4.01, XHTML101, or XHTML1.1 to make the script generate a standalone, standards compliant page.
IP -l
Add the "rel" tag to all of the links to use LightBox for displaying images.
.IP -L
Add the "rel" tag to all of the links to use LightBox groups for displaying images. The group name is based on the gallery title, so this option should only be used with -t "title" or -T. 
.IP -q
Be quiet.  Suppress all output.  By default the script is very willing to give information about what it is doing.
.IP "-r relative directory"
Directory to that the links should be relative to.  Escape spaces with a \e symbol.
.IP "-s source directory"
Use this directory as a source for images.  Images will be copied from here into the directory
.IP "-S small_image_size_in_pixels"
The size of the thumbnail images.  The number is applied to the longest side and is measured in pixels.
.IP "-t title"
Title of the gallery.  Quote when using spaces.
.IP -T
Generate a title for the gallery based on directory name.
.IP -V
Prints the version (revision) number.
.SH EXAMPLES
.IP "gallery -h"
Displays help text.

.IP "gallery -d /tmp/example"
Script assumes that
.I /tmp/example
is a directory full of images.  These images will be manipulated to the right size and placed into folders.  An index page will be created with HTML markup for displaying the images.

.IP "gallery -r /tmp/mySite -d /tmp/mySite/gallery/example"
This will generate a gallery with images that are relative to 
.I /tmp/mySite.

For example:
.I <a href='mySite/gallery/example/IMG_9313.JPG'>...

.IP "gallery -s ~/Photos/Album/My\e Web\e Album -d /web/mySite/gallery/example"
This will copy all of the images from
.I ~/Photos/Album/My\e Web\e Album
to the
.I /web/mySite/gallery/example
directory.  It will then process the script as usual.  This option still works with the -r option.

.IP "gallery -t \(dqMy Gallery\(dq -d /tmp/example"
Will give the gallery the title 
.B My Gallery.

.IP "gallery -T -d /tmp/example"
The gallery will get the title 
.B example.

.IP "gallery -LT -d /tmp/example"
The gallery will get the title 
.B example
and will be grouped for viewing with LightBox using the same title.  For example: 
.B <a href='example/IMG_9317.JPG' rel='lightbox[example]'>

.IP "gallery -e \(dqhtml\(dq -d /tmp/example"
Will create the file
.IR /tmp/example/index.html .
The default is index.php.

.IP "gallery -c 5 -d /tmp/example"
Will create a gallery with five columns.

.IP "gallery -Tl -r /Users/benlavery/Sites/hashbang0 -d /Users/benlavery/Sites/hashbang0/gallery/My\e Holiday"
This is an example of the command I use for creating galleries for my own site.  Notice how spaces in the path are escaped with a \e.
.SH BUGS
No bugs that I know of!  Please report any bugs to author!
.SH AUTHOR
Ben Lavery <ben.lavery@gmail.com>
.SH "SEE ALSO"
.BR sips (1),
.BR convert (1),
.BR identify (1),

sigc_script/perl/gallery.pl
#!/usr/bin/env perl

# Author Ben Lavery
# Contact the author: ben.lavery@gmail.com
# Author's website: hashbang0.com

# First created on 15/05/2010
# Perl script created on 08/07/2010
# Last edited on 12/07/2010

# Project site:
# http://code.google.com/p/simple-image-gallery-creator/

###############################################################################
################################ Licence ######################################
###############################################################################
# This code is licensed under the new BSD licence:
#
#Copyright (c) 2010, Ben Lavery
#All rights reserved.
#
#Redistribution and use in source and binary forms, with or without
#modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright notice, 
#      this list of conditions and the following disclaimer.
#    * Redistributions in binary form must reproduce the above copyright notice,
#      this list of conditions and the following disclaimer in the documentation
#      and/or other materials provided with the distribution.
#
#THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
#ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
#WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
#DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
#FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
#DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
#SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
#CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, 
#OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
#OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
###############################################################################


###############################################################################
################################ Use modules ##################################
###############################################################################
#use warnings;
use 5.010;
use Getopt::Long qw(:config no_ignore_case bundling);
###############################################################################


###############################################################################
################################ "Global" variables ###########################
###############################################################################
#### SCRIPT VERSION ####
$REVISION = '0.10';

#### FILE SETTINGS ####
$INDEX_EXTENTION = "php";

#### GALLERY SETTINGS ####
$CAPTIONS = 0;
$TITLE = undef;
$GENERATE_TITLE = 0;
$COLUMNS = 4;
$LIGHTBOX = 0;
$LIGHTBOX_GROUP = 0;
$DESCRIPTION = undef;

##### HTML SETTINGS ####
$FULL_HTML = 0;
$HTML4 = "HTML4.01";
$XHTML1 = "XHTML1.0";
$XHTML11 = "XHTML1.1";

#### IMAGE SETTINGS ####
$THUMB_PIXELS = 128;
$BIG_PIXELS = 640;
@IMAGES = undef;

#### SCRIPT SETTINGS ####
$QUIET = 0;
$EXIT_STATUS = 0;
$EXIT_REASON = undef;
$SEARCH_PATH = undef;
$HELP = 0;
$VERSION = 0;

#### DIRECTORIES & FILES ####
$ROOT = undef;
$THUMBS = undef;
$BIGS = undef;
$INDEX_FILE = undef;
$INDEX = undef;
$NEWPATH = undef;
$SOURCE = undef;
$RELATIVE = undef;
$DIRECTORY = undef;
$DESCRIPTION_FILE = undef;
$STORE = undef;
###############################################################################


###############################################################################
################################ Cleanup Function #############################
###############################################################################
sub cleanup{
	#Do nothing so far
}

###############################################################################


###############################################################################
################################ Help Function ################################
###############################################################################
sub usage(){
	say "DON'T PANIC!";
}

###############################################################################


###############################################################################
#/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\MAIN/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/#
###############################################################################
#				ooo        ooooo       .o.       ooooo ooooo      ooo         #
#				`88.       .888'      .888.      `888' `888b.     `8'         #
#				 888b     d'888      .8"888.      888   8 `88b.    8          #
#				 8 Y88. .P  888     .8' `888.     888   8   `88b.  8          #
#				 8  `888'   888    .88ooo8888.    888   8     `88b.8          #
#				 8    Y     888   .8'     `888.   888   8       `888          #
#				o8o        o888o o88o     o8888o o888o o8o        `8          #
###############################################################################
#/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\MAIN/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/#
###############################################################################

############ Collect arguments ############

	GetOptions(
		'B|big_size=s' => \$BIG_PIXELS,
		'c|columns:i' => \$COLUMNS,
		'C|captions' => \$CAPTIONS,
		'd|gallery_directory=s' => \$DIRECTORY,
		'D|description_location=s' => \$DESCRIPTION,
		'description_file=s' => \$DESCRIPTION_FILE,
		'e|extention=s' => \$INDEX_EXTENTION,
		'h|help|?' => \$HELP,
		'H|standalone_html_version=s' => \$FULL_HTML,
		'l' => \$LIGHTBOX,
		'L' => \$LIGHTBOX_GROUP,
		'q|quiet' => \$QUIET,
		'r|relative_to=s' => \$RELATIVE,
		's|source_directory=s' => \$SOURCE,
		'S|thumb_size=i' => \$THUMB_PIXELS,
		't=s' => \$TITLE,
		'T' => \$GENERATE_TITLE,
		'V|version' => \$VERSION
	);


############ Check arguments ############
&usage() and exit if $HELP;
say "You are running v$REVISION" and exit if $VERSION;

# Check the gallery directory first.  If it's not there, why continue?
die "The gallery directory was no specified.  Stopped" unless $DIRECTORY;
die "The gallery directory ($DIRECTORY) was not found, or it is not a directory.  Stopped" unless -d $DIRECTORY;

# Shave off any slashes at the end of the path
$DIRECTORY =~  s/\/$//;
$SOURCE =~  s/\/$// if $SOURCE;
$RELATIVE =~  s/\/$// if $RELATIVE;

# If user specified a title and told the script to generate a title 
# just take the title the user gave us and ignor the generation.
if($TITLE && $GENERATE_TITLE){
	$GENERATE_TITLE = 0;
}

# If LIGHTBOX_GROUP is set, we will group images by gallery name
# But only if TITLE has a non-zero value or a title is going to be generated
if($LIGHTBOX_GROUP){
	die "You have requested to group lightbox images, but have not provided a title or requested one to be auto-generated.  Stopped" unless $TITLE or $GENERATE_TITLE;
}

# Check user specified a correct HTML version
if($FULL_HTML){
	die "You have specified an unknown HTML version.  Valid versions are: $HTML4, $XHTML1, and $XHTML11.  Stopped" unless $FULL_HTML eq $HTML4 or $FULL_HTML eq $XHTML1 or $FULL_HTML eq $XHTML11;
}

# If user specifies full html, AND they ask for Lightbox, we'll throw an error.
# This combination is CURRENTLY unsupported
die "You have asked the script to create a standalone gallery page WITH lightbox capabilities.  This combination is UNSUPPORTED at this time.  Stopped" if $FULL_HTML and ($LIGHTBOX or $LIGHTBOX_GROUP);

# Check user specified a correct position for the decription
if($DESCRIPTION){
	die "There is no description file.  Please create the description file under $DIRECTORY/description.txt.  Stopped" unless -f $DIRECTORY . "/description.txt";
	die "You have specified an unknown location for the description text.  Valid locations are 'above' and 'below'.  Stopped" unless $DESCRIPTION eq "above" or $DESCRIPTION eq "below";
}

############ Check directories ############
## Check SOURCE directory if specified
if($SOURCE){
	die "Source directory and destination directory are the same.  Stopped" if $SOURCE eq $DIRECTORY;
	die "Source directory ($SOURCE) does not exist.  Stopped" unless -d $SOURCE;
}

## Check RELATIVE directory if specified
if($RELATIVE){
	die "Relavite directory ($RELATIVE) does not exist.  Stopped" unless -d $RELATIVE;
}

############ Populate variables with stuff ############
$ROOT = "$DIRECTORY";
$THUMBS = "$ROOT/thumbs";
$BIGS = "$ROOT/bigs";
$STORE = "$ROOT/.store";
$INDEX_FILE = "index.$INDEX_EXTENTION";
$INDEX = "$ROOT/$INDEX_FILE";
$DESCRIPTION_FILE = "$ROOT/description.txt" unless $DESCRIPTION_FILE;

############ Create directories ############
mkdir $THUMBS;
mkdir $BIGS;
mkdir $STORE;

############ Sort images into directories ############


############ Resize images ############
############ Generate index file ############
############ Print index file ############
############ Cleanup ############